core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -248,34 +248,25 @@ sealed abstract class IO[E, A] { self =>
   final def ensuring(finalizer: Infallible[Unit]): IO[E, A] =
     new IO.Ensuring(self, finalizer)
 
-  /**
-   * Executes the release action only if there was an error.
+  /**	
+   * Executes the release action only if there was an error.	
    */
   final def bracketOnError[B](release: A => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
-    bracket0(
-      (r: ExitResult[E, B], a: A) =>
-        r match {
-          case ExitResult.Failed(_)     => release(a)
-          case ExitResult.Terminated(_) => release(a)
-          case _                        => IO.unit
-      }
-    )(use)
+    IO.bracket0(this)((a: A, _: Option[Either[E, B]]) => release(a))(use)
```

Comment:
This is not quite correct — we want to execute release if and only if the option is `None` (interrupted) or `Some(Left(...))` (error), but not if it's `Some(Right(...))` (value).

Reply:
Correct, I wanted to go too fast 😲